File: testdata/fixtures/go/vulnerable/header_injection.go
package handler

import (
	"net/http"
)

func setCustomHeader(w http.ResponseWriter, r *http.Request) {
	// VULNERABLE: setting header from user input without CRLF sanitization
	customVal := r.URL.Query().Get("val")
	w.Header().Set("X-Custom", customVal)

	// VULNERABLE: directly using request parameter in header
	w.Header().Set("X-Forwarded-User", r.FormValue("user"))
}

File: batou-rules/testdata/fixtures/javascript/vulnerable/header_injection.ts
import express from 'express';

const app = express();

app.get('/api/data', (req, res) => {
  // VULNERABLE: setting header from req.query without CRLF sanitization
  res.setHeader('X-Request-Id', req.query.requestId);
  res.set('X-Custom-Header', req.query.custom);
  res.header('X-User-Agent', req.headers['x-forwarded-for']);
  res.json({ ok: true });
});
